src/EventStore.Projections.Core/Services/Interpreted/JintProjectionStateHandler.cs

```diff
@@ -993,9 +987,266 @@ private bool EnsureLinkMetadata(out JsValue value)
+								return true;
+							} else {
+								_parent.PushWriter(new ObjectWriter(current.AsObject(), _parent));
+								_position++;
+								return true;
+							}
+						}
+						SerializePrimitive(current, writer);
+					}
+					writer.WriteEndArray();
+					return _parent.PopWriter();
+				}
+			}
+
+			class ObjectWriter : IWriter {
+				private readonly IEnumerator<KeyValuePair<JsValue, PropertyDescriptor>> _iterator;
+				private readonly Serializer _parent;
+				private bool _started;
+
+				public ObjectWriter(ObjectInstance instance, Serializer parent) {
+					_iterator = instance.GetOwnProperties().GetEnumerator();
+					_parent = parent;
+				}
+
+				public bool Write(Utf8JsonWriter writer) {
+					if (!_started) {
+						writer.WriteStartObject();
+						_started = true;
+					}
+					while (_iterator.MoveNext()) {
+						var (name, propertyDescriptor) = _iterator.Current;
+						var value = propertyDescriptor.Value;
+						_parent.WriteMaybeCachedPropertyName(name.AsString());
+						if (value.Type == Types.Object) {
+							if (value is ArrayInstance ai) {
+								_parent.PushWriter(new ArrayWriter(ai, _parent));
+								return true;
+							} else {
+								_parent.PushWriter(new ObjectWriter(value.AsObject(), _parent));
+								return true;
+							}
+						} else {
+							SerializePrimitive(value, writer);
+						}
+
+					}
+
+					writer.WriteEndObject();
+					return _parent.PopWriter();
+				}
+			}
+
+			[MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+			private void WriteMaybeCachedPropertyName(string name) {
+				if (_knownPropertyNames.TryGetValue(name, out var propertyName)) {
+					_writer.WritePropertyName(propertyName);
+					return;
+				}
+				if (_knownPropertyNames.Count < 1000) {
+					_knownPropertyNames.Add(name, JsonEncodedText.Encode(name));
 				}
+				_writer.WritePropertyName(name);
 			}
+
+			class PrimitiveWriter : IWriter {
+				private readonly JsValue _value;
+
+				public PrimitiveWriter(JsValue value) {
+					_value = value;
+				}
+
+				public bool Write(Utf8JsonWriter writer) {
+					SerializePrimitive(_value, writer);
+					return false;
+				}
+			}
+			private class CompletedWriter : IWriter {
+				public bool Write(Utf8JsonWriter writer) {
+					return false;
+				}
+			}
+
+			[MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+			static void SerializePrimitive(JsValue value, Utf8JsonWriter writer) {
+
+				switch (value.Type) {
+					case Types.Null:
+					case Types.Undefined:
+					case Types.None:
+						writer.WriteNullValue();
+						break;
+					case Types.Boolean:
+						if (ReferenceEquals(value, JsBoolean.False))
+							writer.WriteBooleanValue(false);
+						else
+							writer.WriteBooleanValue(true);
+						break;
+					case Types.Number:
+						writer.WriteNumberValue(value.AsNumber());
+						break;
+					case Types.BigInt:
+						writer.WriteStringValue(value.ToString());
```

Comment:
is this a limitation of the writer, not to write bigints as numbers? or maybe a js convention - either way perhaps a comment so it's clear its intentional. otherwise its a bit surprising to see

Reply:
for the js bigint type, it is safer to write them as strings cos json doesn't have bigints, and bigint is primarily for numbers that don't fit in the number type